### PR TITLE
chore(users): removed UsersTable class

### PR DESCRIPTION
**Pull request by jeabakker**
moved function to better locations

ref #14095

### CODE REVIEW

File: actions/login.php [hunk 1]
```diff
@@ -15,20 +15,19 @@
...
-	return get_user_by_username($username);
+$user = elgg_call(ELGG_SHOW_DISABLED_ENTITIES, function() use (&$username) {
+	// check if logging in with email address
+	if (elgg_strpos($username, '@') !== false) {
+		$user = elgg_get_user_by_email($username);
+		if ($user instanceof \ElggUser) {
+			$username = $user->username;
+		}
+		
+		return $user;
```
Comment (jdalsem) [inline actions/login.php:27]: only return user if instance of user so it falls back to username fetch

File: actions/user/requestnewpassword.php [hunk 2]
```diff
@@ -9,11 +9,12 @@
 }
 
 // allow email addresses
-if (elgg_strpos($username, '@') !== false && ($users = get_user_by_email($username))) {
-	$username = $users[0]->username;
+if (elgg_strpos($username, '@') !== false) {
+	$user = elgg_get_user_by_email($username);
+} else {
```
Comment (jdalsem) [inline actions/user/requestnewpassword.php:14]: not correct logic... username fetch should always happen if not found by email

File: engine/lib/users.php [hunk 3]
```diff
@@ -28,25 +28,72 @@ function get_user(int $guid): ?\ElggUser {
 }
 
 /**
- * Get user by username
+ * Get an user by username
```
Comment (jdalsem) [inline engine/lib/users.php:31]: a user..

File: engine/lib/users.php [hunk 4]
```diff
@@ -28,25 +28,72 @@ function get_user(int $guid): ?\ElggUser {
...
+		],
+		'limit' => 1,
+	]);
+	
+	return $users ? $users[0] : null;
 }
 
 /**
- * Get an array of users from an email address
+ * Get an user from an email address
```
Comment (jdalsem) [inline engine/lib/users.php:72]: a user

File: engine/classes/Elgg/Router/Middleware/MaintenanceGatekeeper.php [hunk 5]
```diff
@@ -60,13 +60,10 @@ protected function isAllowedAction(Request $request): bool {
...
 		$username = (string) $request->getParam('username');
 
-		$user = get_user_by_username($username);
-
-		if (!$user instanceof \ElggUser) {
-			$users = get_user_by_email($username);
-			if (!empty($users)) {
-				$user = $users[0];
-			}
+		if (elgg_strpos('@', $username) !== false) {
```
Comment (jdalsem) [inline engine/classes/Elgg/Router/Middleware/MaintenanceGatekeeper.php:63]: change pattern

File: mod/web_services/classes/Elgg/WebServices/ApiMethods/AuthGetToken.php [hunk 6]
```diff
@@ -29,14 +29,16 @@ class AuthGetToken {
 	public function __invoke(string $username, string $password) {
 		// check if username is an email address
 		if (elgg_is_valid_email($username)) {
-			$users = get_user_by_email($username);
+			$user = elgg_get_user_by_email($username);
```
Comment (jdalsem) [inline mod/web_services/classes/Elgg/WebServices/ApiMethods/AuthGetToken.php:32]: pattern

**Review by jdalsem — CHANGES_REQUESTED** [review]

**Review by jdalsem — CHANGES_REQUESTED** [review]

File: engine/lib/users.php [hunk 7]
```diff
@@ -28,25 +28,80 @@ function get_user(int $guid): ?\ElggUser {
...
+ * @since 5.0
  */
-function get_user_by_username(string $username): ?\ElggUser {
-	return _elgg_services()->usersTable->getByUsername($username);
+function elgg_get_user_by_username(string $username, bool $try_email = false): ?\ElggUser {
+	if (empty($username)) {
+		return null;
+	}
+	
+	if ($try_email && elgg_is_valid_email($username)) {
```
Comment (jdalsem) [inline engine/lib/users.php:44]: maybe move this after the elgg_get_entities from username... the login action will allow try email, but mostly the login action is by username. This introduces an extra query... moving it after the username query will have this more like a fallback, as the primary login logic fails...

File: engine/classes/Elgg/Router/Route.php [hunk 8]
```diff
@@ -109,7 +109,7 @@ public function resolvePageOwner() {
 				
 				$username = elgg_extract('username', $params);
 				if ($username) {
-					return get_user_by_username($username) ?: null;
+					return elgg_get_user_by_username($username) ?: null;
```
Comment (jdalsem) [inline engine/classes/Elgg/Router/Route.php:112]: no need for the ?:null as the function will return null

File: engine/lib/users.php (hunk 7 above)
Comment (jeabakker) [reply engine/lib/users.php:44]: Not going to change, the logic always was:
- if email try email
- else fallback to username

This way the logic stays intact

**Review by jdalsem — APPROVED** [review]

**Pull request merged**